### DISCUSSION TITLE
Problems enabling Two Factor Auth

### DISCUSSION THREAD

**Discussion by stfnx** [pallets-eco/flask-security#963] [Q&A] [answered]
Hi,
I've implemented this library and everything works fine so far. Now I tried to add Two Factor Auth. For that I added some config options:
```py
SECURITY_TWO_FACTOR = True
SECURITY_TWO_FACTOR_ENABLED_METHODS = ['email', 'authenticator']
SECURITY_TOTP_ISSUER = 'myapp'
SECURITY_TOTP_SECRETS = {'1': os.getenv("SECURITY_TOTP_SECRET")}
```

My user model looks like this:
```py
from flask_security.models import fsqla_v3 as fsqla

class User(db.Model, fsqla.FsUserMixin):
    first_name = db.Column(db.String(255))
    last_name = db.Column(db.String(255))
    image_filename = db.Column(db.String(255))
```

And the database looks like this:
![image](https://github.com/Flask-Middleware/flask-security/assets/66280727/a622ebc4-e144-41b3-bc74-9d4d394739cb)

It seems that all the necessary columns are already there. Provided through `fsqla_v3.FsUserMixin`.

When I try to access any route I get the following error:
```
sqlalchemy.exc.InvalidRequestError: When initializing mapper Mapper[User(user)], expression 'WebAuthn' failed to locate a name ('WebAuthn'). If this is a class name, consider adding this relationship() to the <class 'webdev_template.blueprints.auth.models.User'> class after both dependent classes have been defined.
```

I don't plan to use WebAuthn yet. The config option `SECURITY_WEBAUTHN` should be `False` by default and I havn't enabled it.

What is going on here? Do I need to create a WebAuthn db table nonetheless?

**Comment by stfnx** [comment, marked as answer]
Yeah... seems like I have to. Adding the table like so worked:
```
class WebAuthn(db.Model, fsqla.FsWebAuthnMixin):
    pass
```

**Comment by jwag956** [comment]
You are correct - the fsqla are designed to be 'batteries included' and to help applications future-proof themselves w/o requiring DB schema updates. I noticed that the quickstart wasn't updated to show this - I'll do that.
Thanks.

**Reply by dodumosu** [reply]
I had a bit of a nasty surprise recently. I've been used to subclassing `RoleMixin` and `UserMixin` for my auth model classes, but I recently started using the `FsModels` classes on a project, and made the mistake (something I realized later) of selecting the `mfa` extra for 2FA. However, I was unable to create users at the CLI, and kept getting an error about the `WebAuthn` model being missing. I thought it was due to subclassing the `FsModels` classes, but after digging through the source, I found that the very presence of the `webauth` package immediately enables the utility of the `WebAuthn` model class and relationship, even if one isn't actually interested in it per se.

I think perhaps that should be moved to a configuration setting, rather than being a hard dependency on an installed package.
